lib/handler/mruby.c

```diff
@@ -327,38 +330,65 @@ static mrb_value build_constants(mrb_state *mrb, const char *server_name, size_t
     return ary;
 }
 
-static void on_context_init(h2o_handler_t *_handler, h2o_context_t *ctx)
+static h2o_mruby_shared_context_t *create_shared_context(h2o_context_t *ctx)
 {
-    h2o_mruby_handler_t *handler = (void *)_handler;
-    h2o_mruby_context_t *handler_ctx = h2o_mem_alloc(sizeof(*handler_ctx));
-
-    handler_ctx->handler = handler;
-
     /* init mruby in every thread */
-    if ((handler_ctx->mrb = mrb_open()) == NULL) {
+    h2o_mruby_shared_context_t *shared_ctx = h2o_mem_alloc(sizeof(*shared_ctx));
+    if ((shared_ctx->mrb = mrb_open()) == NULL) {
         fprintf(stderr, "%s: no memory\n", H2O_MRUBY_MODULE_NAME);
         abort();
     }
-    handler_ctx->constants = build_constants(handler_ctx->mrb, ctx->globalconf->server_name.base, ctx->globalconf->server_name.len);
-    handler_ctx->symbols.sym_call = mrb_intern_lit(handler_ctx->mrb, "call");
-    handler_ctx->symbols.sym_close = mrb_intern_lit(handler_ctx->mrb, "close");
-    handler_ctx->symbols.sym_method = mrb_intern_lit(handler_ctx->mrb, "method");
-    handler_ctx->symbols.sym_headers = mrb_intern_lit(handler_ctx->mrb, "headers");
-    handler_ctx->symbols.sym_body = mrb_intern_lit(handler_ctx->mrb, "body");
-    handler_ctx->symbols.sym_async = mrb_intern_lit(handler_ctx->mrb, "async");
+    h2o_mruby_setup_globals(shared_ctx->mrb);
+    shared_ctx->constants = build_constants(shared_ctx->mrb, ctx->globalconf->server_name.base, ctx->globalconf->server_name.len);
+    shared_ctx->symbols.sym_call = mrb_intern_lit(shared_ctx->mrb, "call");
+    shared_ctx->symbols.sym_close = mrb_intern_lit(shared_ctx->mrb, "close");
+    shared_ctx->symbols.sym_method = mrb_intern_lit(shared_ctx->mrb, "method");
+    shared_ctx->symbols.sym_headers = mrb_intern_lit(shared_ctx->mrb, "headers");
+    shared_ctx->symbols.sym_body = mrb_intern_lit(shared_ctx->mrb, "body");
+    shared_ctx->symbols.sym_async = mrb_intern_lit(shared_ctx->mrb, "async");
+
+    h2o_mruby_send_chunked_init_context(shared_ctx);
+    h2o_mruby_http_request_init_context(shared_ctx);
+
+    return shared_ctx;
+}
 
-    h2o_mruby_send_chunked_init_context(handler_ctx);
-    h2o_mruby_http_request_init_context(handler_ctx);
+static void dispose_storage_item(void *data)
+{
+    if (data == NULL)
+        return;
+    h2o_mruby_shared_context_t *shared_ctx = (h2o_mruby_shared_context_t *)data;
+    mrb_close(shared_ctx->mrb);
```

Comment:
Maybe you need to free `shared_ctx`?

Also, it might be a good idea to rename the function to `dispose_shared_context`, since it disposes the context created by `create_shared_context`.